memory_allocator.c

```diff
@@ -16,9 +16,16 @@
 TAILQ_HEAD(mb_list, mem_block);
 TAILQ_HEAD(mp_arena, mem_arena);
 
+int32_t abs(int32_t x)
+{
+  if (x < 0)
+    return -x;
+  return x;
+}
+
 typedef struct mem_block {
   uint32_t mb_magic; /* if overwritten report a memory corruption error */
-  uint32_t mb_size;
+  int32_t mb_size; /* size < 0 => free, size > 0 => alloc'd */
```

Comment:
Wouldn't you be better off with the opposite encoding - i.e. `/* size > 0 => free, size < 0 => alloc'd */`?

You would have to consider `size < 0` case only for allocated blocks then, instead of using `abs` all the time.